src/concurrency.md

```diff
@@ -232,6 +233,96 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[threadpool-walk]: #threadpool-walk
+<a name="threadpool-walk"></a>
+
+## Calculate SHA1 sum of *.iso files concurrently
+
+[![threadpool-badge]][threadpool] [![num_cpus-badge]][num_cpus] [![walkdir-badge]][walkdir] [![cat-concurrency-badge]][cat-concurrency][![cat-filesystem-badge]][cat-filesystem]
+
+This example uses `threadpool`, `walkdir` and `ring` crates to calculate the SHA1 for every file present in the current directory. A threadpool `pool` is created using the number of cpus present in the system. Then every `entry` returned by `walkdir` is passed into this pool to perform the operations of reading and compute SHA1. At the end the program waits for all jobs to finish. To get better results, compile this program in release mode.
+
+```rust,no_run
+# #[macro_use]
+# extern crate error_chain;
+extern crate walkdir;
+extern crate ring;
+extern crate num_cpus;
+extern crate threadpool;
+
+# error_chain! {
+#     foreign_links {
+#         Io(std::io::Error);
+#     }
+# }
+
+use walkdir::WalkDir;
+use std::fs::File;
+use std::io::BufReader;
+use std::io::prelude::Read;
+use std::path::Path;
+use threadpool::ThreadPool;
+use std::sync::mpsc::channel;
+use ring::digest::{Context, Digest, SHA256};
+
+// Verify the iso extension
+# fn is_iso(entry: &Path) -> bool {
+#     match entry.extension() {
+#         Some(e) if e == "iso"=> true,
+#         Some(_) => false,
+#         None => false,
+#     }
+# }
+
+fn compute_digest<P: AsRef<Path>>(filepath: P) -> Result<(Digest, String)> {
```

Comment:
if you return `Result<(Digest, P)>` you can completely drop the `filename` variable and just return `Ok((context.finish(), filepath))`